Introduce OptimizerInfos + add a test_errors

janeyx99 opened the pull request:
Introduce OptimizerInfos + use them to refactor out the error testing.

Why OptimizerInfos?
- cleaner, easier way to test all configs of optimizers
- would plug in well with devicetype to auto-enable tests for devices like MPS, meta
- would allow for more granular testing. currently, lots of functionality is tested in `_test_basic_cases` and some of that should be broken down more.

What did I do for error testing?
- I moved out some error cases from `_test_basic_cases` into a new test_errors parametrized test.
- The new test has to live in TestOptimRenewed (bikeshedding welcome) because the parametrized tests need to take in device and dtype and hook correctly, and not all tests in TestOptim do that.
- TestOptimRenewed also is migrating to the toplevel test/test_optim.py now because importing TestOptimRenewed does not work (because of test instantiation, TestOptimRenewed gets replaced with TestOptimRenewedDevice for CPU, CUDA, and whatever other device).

Is there any change in test coverage?
- INCREASE: The error case where a single Parameter (vs a container of them) are passed in has now expanded to all optims instead of only LBFGS
- DECREASE: Not much. The only thing is we no longer test two error cases for foreach=True AND foreach=False, which I think is redundant. (Highlighted in comments)

Possible but not urgent next step: test ALL possible error cases by going through all the constructors.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115025
* #115023
* #114802
* #114797
* __->__ #114178

janeyx99 committed:
[init] Add OptimizerInfos

[ghstack-poisoned]

janeyx99 committed:
Update on "[init] Add OptimizerInfos"




[ghstack-poisoned]

janeyx99 committed:
Update on "[init] Add OptimizerInfos"




[ghstack-poisoned]

janeyx99 committed:
Update on "[init] Add OptimizerInfos"




[ghstack-poisoned]

janeyx99 committed:
Update on "[init] Add OptimizerInfos"




[ghstack-poisoned]

janeyx99 commented on test/optim/test_optim.py:
```diff
@@ -1715,8 +1666,7 @@ def test_rmsprop(self):
                 lambda param: RMSprop([param], maximize=True, foreach=foreach)
             )
-            with self.assertRaisesRegex(ValueError, "Invalid momentum value: -1.0"):
-                RMSprop(None, lr=1e-2, momentum=-1.0, foreach=foreach)
```
This is one place where coverage "decreased", since I no longer test this error case for BOTH foreach T and F. Which I believe is unnecessary anyway.

jbschlosser replied:
maybe this is unnecessary, but it's easy enough to do this:
```python
class TestOptimRenewed:
    ...
    @onlyCPU
    @optims([optim for optim in optim_db if optim.optim_error_inputs_func is not None])
    @parametrize("foreach", [False, True])
    def test_errors(self, device, dtype, optim_info, foreach):
        ...
```

janeyx99 replied:
Yea, thanks for the code pointer. I don't believe it is necessary though.

janeyx99 commented on test/optim/test_optim.py:
```diff
@@ -1805,8 +1754,7 @@ def test_rprop(self):
                 )
             )
-            with self.assertRaisesRegex(ValueError, "Invalid eta values: 1.0, 0.5"):
-                Rprop(None, lr=1e-2, etas=(1.0, 0.5), foreach=foreach)
```
And this is the other place where coverage "decreased", since I no longer test this error case for BOTH foreach T and F. Which I believe is unnecessary anyway.

janeyx99 committed:
Update on "Introduce OptimizerInfos + add a test_errors"

Introduce OptimizerInfos + use them to refactor out the error testing.

Why OptimizerInfos?
- cleaner, easier way to test all configs of optimizers
- would plug in well with devicetype to auto-enable tests for devices like MPS, meta
- would allow for more granular testing. currently, lots of functionality is tested in `_test_basic_cases` and some of that should be broken down more.

What did I do for error testing?
- I moved out some error cases from `_test_basic_cases` into a new test_errors parametrized test.
- The new test has to live in TestOptimRenewed (bikeshedding welcome) because the parametrized tests need to take in device and dtype and hook correctly, and not all tests in TestOptim do that.
- TestOptimRenewed also is migrating to the toplevel test/test_optim.py now because importing TestOptimRenewed does not work (because of test instantiation, TestOptimRenewed gets replaced with TestOptimRenewedDevice for CPU, CUDA, and whatever other device).

Is there any change in test coverage?
- INCREASE: The error case where a single Parameter (vs a container of them) are passed in has now expanded to all optims instead of only LBFGS
- DECREASE: Not much. The only thing is we no longer test two error cases for foreach=True AND foreach=False, which I think is redundant. (Highlighted in comments)




[ghstack-poisoned]

janeyx99 committed:
Update on "Introduce OptimizerInfos + add a test_errors"

Introduce OptimizerInfos + use them to refactor out the error testing.

Why OptimizerInfos?
- cleaner, easier way to test all configs of optimizers
- would plug in well with devicetype to auto-enable tests for devices like MPS, meta
- would allow for more granular testing. currently, lots of functionality is tested in `_test_basic_cases` and some of that should be broken down more.

What did I do for error testing?
- I moved out some error cases from `_test_basic_cases` into a new test_errors parametrized test.
- The new test has to live in TestOptimRenewed (bikeshedding welcome) because the parametrized tests need to take in device and dtype and hook correctly, and not all tests in TestOptim do that.
- TestOptimRenewed also is migrating to the toplevel test/test_optim.py now because importing TestOptimRenewed does not work (because of test instantiation, TestOptimRenewed gets replaced with TestOptimRenewedDevice for CPU, CUDA, and whatever other device).

Is there any change in test coverage?
- INCREASE: The error case where a single Parameter (vs a container of them) are passed in has now expanded to all optims instead of only LBFGS
- DECREASE: Not much. The only thing is we no longer test two error cases for foreach=True AND foreach=False, which I think is redundant. (Highlighted in comments)




[ghstack-poisoned]

janeyx99 committed:
Update on "Introduce OptimizerInfos + add a test_errors"

Introduce OptimizerInfos + use them to refactor out the error testing.

Why OptimizerInfos?
- cleaner, easier way to test all configs of optimizers
- would plug in well with devicetype to auto-enable tests for devices like MPS, meta
- would allow for more granular testing. currently, lots of functionality is tested in `_test_basic_cases` and some of that should be broken down more.

What did I do for error testing?
- I moved out some error cases from `_test_basic_cases` into a new test_errors parametrized test.
- The new test has to live in TestOptimRenewed (bikeshedding welcome) because the parametrized tests need to take in device and dtype and hook correctly, and not all tests in TestOptim do that.
- TestOptimRenewed also is migrating to the toplevel test/test_optim.py now because importing TestOptimRenewed does not work (because of test instantiation, TestOptimRenewed gets replaced with TestOptimRenewedDevice for CPU, CUDA, and whatever other device).

Is there any change in test coverage?
- INCREASE: The error case where a single Parameter (vs a container of them) are passed in has now expanded to all optims instead of only LBFGS
- DECREASE: Not much. The only thing is we no longer test two error cases for foreach=True AND foreach=False, which I think is redundant. (Highlighted in comments)




[ghstack-poisoned]

janeyx99 committed:
Update on "Introduce OptimizerInfos + add a test_errors"

Introduce OptimizerInfos + use them to refactor out the error testing.

Why OptimizerInfos?
- cleaner, easier way to test all configs of optimizers
- would plug in well with devicetype to auto-enable tests for devices like MPS, meta
- would allow for more granular testing. currently, lots of functionality is tested in `_test_basic_cases` and some of that should be broken down more.

What did I do for error testing?
- I moved out some error cases from `_test_basic_cases` into a new test_errors parametrized test.
- The new test has to live in TestOptimRenewed (bikeshedding welcome) because the parametrized tests need to take in device and dtype and hook correctly, and not all tests in TestOptim do that.
- TestOptimRenewed also is migrating to the toplevel test/test_optim.py now because importing TestOptimRenewed does not work (because of test instantiation, TestOptimRenewed gets replaced with TestOptimRenewedDevice for CPU, CUDA, and whatever other device).

Is there any change in test coverage?
- INCREASE: The error case where a single Parameter (vs a container of them) are passed in has now expanded to all optims instead of only LBFGS
- DECREASE: Not much. The only thing is we no longer test two error cases for foreach=True AND foreach=False, which I think is redundant. (Highlighted in comments)




[ghstack-poisoned]

janeyx99 committed:
Update on "Introduce OptimizerInfos + add a test_errors"

Introduce OptimizerInfos + use them to refactor out the error testing.

Why OptimizerInfos?
- cleaner, easier way to test all configs of optimizers
- would plug in well with devicetype to auto-enable tests for devices like MPS, meta
- would allow for more granular testing. currently, lots of functionality is tested in `_test_basic_cases` and some of that should be broken down more.

What did I do for error testing?
- I moved out some error cases from `_test_basic_cases` into a new test_errors parametrized test.
- The new test has to live in TestOptimRenewed (bikeshedding welcome) because the parametrized tests need to take in device and dtype and hook correctly, and not all tests in TestOptim do that.
- TestOptimRenewed also is migrating to the toplevel test/test_optim.py now because importing TestOptimRenewed does not work (because of test instantiation, TestOptimRenewed gets replaced with TestOptimRenewedDevice for CPU, CUDA, and whatever other device).

Is there any change in test coverage?
- INCREASE: The error case where a single Parameter (vs a container of them) are passed in has now expanded to all optims instead of only LBFGS
- DECREASE: Not much. The only thing is we no longer test two error cases for foreach=True AND foreach=False, which I think is redundant. (Highlighted in comments)

Possible but not urgent next step: test ALL possible error cases by going through all the constructors.




[ghstack-poisoned]

albanD commented on torch/testing/_internal/common_optimizers.py:
```diff
@@ -0,0 +1,626 @@
+        self,
+        optimizer_error_input,
+        *,
+        error_on=OptimizerErrorEnum.CONSTRUCTION_ERROR,
```
Do you expect these to be default being used? If not we can just remove the default values here.

janeyx99 replied:
Yes, most of them are on construction

albanD replied:
Is it true as well for the err type and regex below?

albanD commented on torch/testing/_internal/common_optimizers.py:
```diff
@@ -0,0 +1,626 @@
+        optim_inputs_func,
+        # Implementation specific kwargs the optimizer supports, e.g., fused, foreach, differentiable
+        # We consider capturable to be a base constructor flag since it is implemented across the board.
+        supported_impls: Tuple[str] = ("foreach", "differentiable"),
```
It would be great for the comment above to mention every valid value for this flag.

janeyx99 replied:
wait it does do that

albanD replied:
"kwargs the optimizer supports, e.g., fused, foreach, differentiable"
Is not the same as
"kwargs the optimizer supports, it must be a subset of fused, foreach, differentiable"

albanD commented on torch/testing/_internal/common_optimizers.py:
```diff
@@ -0,0 +1,626 @@
+        optim_inputs_func=None,
+        optim_error_inputs_func=optim_error_inputs_func_sgd,
+        supported_impls=("foreach", "differentiable"),
+        supports_sparse_on=("cpu", "cuda"),
```
Given that the test we have only runs on cpu, the cuda here does nothing in this PR and will be used in a follow above I guess?

janeyx99 replied:
yes! :D

albanD reviewed (approved):
Sounds good!

janeyx99 commented:
@pytorchbot merge

let's see if the 3 days ago is recent enough

janeyx99 commented:
@pytorchbot merge
